File: gateway/internal/models/provider_prefs.go
// provider_prefs.go — per-model OpenRouter provider-routing preferences
// (quick 260830-o2j).
//
// A model_aliases row may carry a `provider_prefs` JSON object that the
// openrouter-chat director injects VERBATIM as the request's `provider`
// field. The schema mirrors OpenRouter's documented provider-routing object
// (https://openrouter.ai/docs/features/provider-routing):
//
//	only / order / ignore                 []string provider slugs
//	allow_fallbacks / require_parameters  bool
//	zdr / enforce_distillable_text        bool
//	data_collection                       "allow" | "deny"
//	quantizations                         []string (int4 … bf16 … unknown)
//	sort                                  "price"|"throughput"|"latency" | {by, partition}
//	max_price                             {prompt, completion, request, image} $/M (>= 0)
//	preferred_min_throughput              number | {p50,p75,p90,p99}
//	preferred_max_latency                 number | {p50,p75,p90,p99}
//
// ValidateProviderPrefs is the SINGLE write-side gate: gatewayctl and the
// admin API both call it before persisting, so the director can trust the
// stored bytes and inject them without re-validating on the hot path.
package models

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"
)

// ProviderPrefsMaxBytes caps the stored JSON so a stray paste cannot balloon
// the resolver cache or the per-request body rewrite.
const ProviderPrefsMaxBytes = 4096

var (
	providerQuantizations = map[string]bool{
		"int4": true, "int8": true, "fp4": true, "mxfp4": true, "nvfp4": true,
		"fp6": true, "fp8": true, "mxfp8": true, "fp16": true, "bf16": true,
		"fp32": true, "unknown": true,
	}
	providerSortBy        = map[string]bool{"price": true, "throughput": true, "latency": true}
	providerSortPartition = map[string]bool{"model": true, "none": true}
	providerPercentiles   = map[string]bool{"p50": true, "p75": true, "p90": true, "p99": true}
	providerDataCollect   = map[string]bool{"allow": true, "deny": true}
)

// ProviderPrefs is the typed mirror of the OpenRouter `provider` object.
// Pointer/omitempty fields so the canonical re-encode only carries what the
// operator set — OpenRouter applies its own defaults for absent keys.
type ProviderPrefs struct {
	Only                   []string          `json:"only,omitempty"`
	Order                  []string          `json:"order,omitempty"`
	Ignore                 []string          `json:"ignore,omitempty"`
	AllowFallbacks         *bool             `json:"allow_fallbacks,omitempty"`
	RequireParameters      *bool             `json:"require_parameters,omitempty"`
	DataCollection         string            `json:"data_collection,omitempty"`
	ZDR                    *bool             `json:"zdr,omitempty"`
	EnforceDistillableText *bool             `json:"enforce_distillable_text,omitempty"`
	Quantizations          []string          `json:"quantizations,omitempty"`
	Sort                   json.RawMessage   `json:"sort,omitempty"`
	MaxPrice               *ProviderMaxPrice `json:"max_price,omitempty"`
	PreferredMinThroughput json.RawMessage   `json:"preferred_min_throughput,omitempty"`
	PreferredMaxLatency    json.RawMessage   `json:"preferred_max_latency,omitempty"`
}

// ProviderMaxPrice is the `max_price` sub-object ($ per million tokens).
type ProviderMaxPrice struct {
	Prompt     *float64 `json:"prompt,omitempty"`
	Completion *float64 `json:"completion,omitempty"`
	Request    *float64 `json:"request,omitempty"`
	Image      *float64 `json:"image,omitempty"`
}

// ValidateProviderPrefs parses raw against the OpenRouter schema (unknown
// keys rejected) and returns the CANONICAL re-encoded bytes to store. An
// empty object `{}` is rejected — callers should store NULL instead.
func ValidateProviderPrefs(raw []byte) ([]byte, error) {
	trimmed := bytes.TrimSpace(raw)
	if len(trimmed) == 0 {
		return nil, errors.New("provider_prefs must not be empty (use NULL to clear)")
	}
	if len(trimmed) > ProviderPrefsMaxBytes {
		return nil, fmt.Errorf("provider_prefs exceeds %d bytes", ProviderPrefsMaxBytes)
	}
	dec := json.NewDecoder(bytes.NewReader(trimmed))
	dec.DisallowUnknownFields()
	var p ProviderPrefs
	if err := dec.Decode(&p); err != nil {
		return nil, fmt.Errorf("provider_prefs: %w", err)
	}
	if dec.More() {
		return nil, errors.New("provider_prefs: trailing data after JSON object")
	}
	if err := p.validate(); err != nil {
		return nil, err
	}
	out, err := json.Marshal(p)
	if err != nil {
		return nil, err
	}
	if string(out) == "{}" {
		return nil, errors.New("provider_prefs must set at least one field (use NULL to clear)")
	}
	return out, nil
}

func (p *ProviderPrefs) validate() error {
	for name, list := range map[string][]string{"only": p.Only, "order": p.Order, "ignore": p.Ignore} {
		if err := validateSlugList(name, list, nil); err != nil {
			return err
		}
	}
	if err := validateSlugList("quantizations", p.Quantizations, providerQuantizations); err != nil {
		return err
	}
	if p.DataCollection != "" && !providerDataCollect[p.DataCollection] {
		return fmt.Errorf("provider_prefs.data_collection must be \"allow\" or \"deny\", got %q", p.DataCollection)
	}
	if p.MaxPrice != nil {
		for name, v := range map[string]*float64{
			"prompt": p.MaxPrice.Prompt, "completion": p.MaxPrice.Completion,
			"request": p.MaxPrice.Request, "image": p.MaxPrice.Image,
		} {
			if v != nil && *v < 0 {
				return fmt.Errorf("provider_prefs.max_price.%s must be >= 0", name)
			}
		}
		if p.MaxPrice.Prompt == nil && p.MaxPrice.Completion == nil && p.MaxPrice.Request == nil && p.MaxPrice.Image == nil {
			return errors.New("provider_prefs.max_price must set at least one of prompt/completion/request/image")
		}
	}
	if len(p.Sort) > 0 {
		canon, err := validateSort(p.Sort)
		if err != nil {
			return err
		}
		p.Sort = canon
	}
	for name, raw := range map[string]*json.RawMessage{
		"preferred_min_throughput": &p.PreferredMinThroughput,
		"preferred_max_latency":    &p.PreferredMaxLatency,
	} {
		if len(*raw) == 0 {
			continue
		}
		canon, err := validateThreshold(name, *raw)
		if err != nil {
			return err
		}
		*raw = canon
	}
	return nil
}

func validateSlugList(name string, list []string, allowed map[string]bool) error {
	if list == nil {
		return nil
	}
	if len(list) == 0 {
		return fmt.Errorf("provider_prefs.%s must not be an empty list (omit the key instead)", name)
	}
	if len(list) > 32 {
		return fmt.Errorf("provider_prefs.%s has too many entries (max 32)", name)
	}
	for _, s := range list {
		if s == "" || strings.TrimSpace(s) != s || strings.ContainsAny(s, " \t\r\n") {
			return fmt.Errorf("provider_prefs.%s entries must be non-empty slugs without whitespace, got %q", name, s)
		}
		if len(s) > 64 {
			return fmt.Errorf("provider_prefs.%s entry %q exceeds 64 chars", name, s)
		}
		if allowed != nil && !allowed[s] {
			keys := make([]string, 0, len(allowed))
			for k := range allowed {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			return fmt.Errorf("provider_prefs.%s: %q is not one of %s", name, s, strings.Join(keys, ","))
		}
	}
	return nil
}

// validateSort accepts "price"|"throughput"|"latency" or {by, partition?}.
func validateSort(raw json.RawMessage) (json.RawMessage, error) {
	var s string
	if err := json.Unmarshal(raw, &s); err == nil {
		if !providerSortBy[s] {
			return nil, fmt.Errorf("provider_prefs.sort must be price|throughput|latency, got %q", s)
		}
		return json.Marshal(s)
	}
	var obj struct {
		By        string `json:"by"`
		Partition string `json:"partition,omitempty"`
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&obj); err != nil {
		return nil, fmt.Errorf("provider_prefs.sort: %w", err)
	}
	if !providerSortBy[obj.By] {
		return nil, fmt.Errorf("provider_prefs.sort.by must be price|throughput|latency, got %q", obj.By)
	}
	if obj.Partition != "" && !providerSortPartition[obj.Partition] {
		return nil, fmt.Errorf("provider_prefs.sort.partition must be model|none, got %q", obj.Partition)
	}
	return json.Marshal(obj)
}

// validateThreshold accepts a non-negative number or an object keyed by
// p50/p75/p90/p99 with non-negative numeric values.
func validateThreshold(name string, raw json.RawMessage) (json.RawMessage, error) {
	var n float64
	if err := json.Unmarshal(raw, &n); err == nil {
		if n < 0 {
			return nil, fmt.Errorf("provider_prefs.%s must be >= 0", name)
		}
		return json.Marshal(n)
	}
	var obj map[string]float64
	if err := json.Unmarshal(raw, &obj); err != nil {
		return nil, fmt.Errorf("provider_prefs.%s must be a number or {p50|p75|p90|p99: number}: %w", name, err)
	}
	if len(obj) == 0 {
		return nil, fmt.Errorf("provider_prefs.%s object must not be empty", name)
	}
	for k, v := range obj {
		if !providerPercentiles[k] {
			return nil, fmt.Errorf("provider_prefs.%s: unknown percentile %q (use p50/p75/p90/p99)", name, k)
		}
		if v < 0 {
			return nil, fmt.Errorf("provider_prefs.%s.%s must be >= 0", name, k)
		}
	}
	return json.Marshal(obj)
}

// ProviderPrefs returns the stored provider object for (alias, upstream) or
// nil when the row has none. The returned slice is a copy of the cached
// bytes so callers may not mutate the cache.
func (r *Resolver) ProviderPrefs(alias, upstream string) []byte {
	r.mu.RLock()
	raw, ok := r.prefs[aliasKey{Alias: alias, Upstream: upstream}]
	r.mu.RUnlock()
	if !ok || len(raw) == 0 {
		return nil
	}
	out := make([]byte, len(raw))
	copy(out, raw)
	return out
}

File: dashboard/src/lib/provider-prefs.ts
/**
 * OpenRouter provider-routing preferences — shared TS mirror of the gateway's
 * `models.ValidateProviderPrefs` (quick 260830-o2j).
 *
 * Used by BOTH the client editor (instant feedback) and the owner server
 * actions (defense-in-depth before the gateway call; the gateway re-validates
 * anyway — it is the authority). Schema follows
 * https://openrouter.ai/docs/features/provider-routing verbatim.
 *
 * Precedence at request time (decided 2026-08-30): tenant.provider_prefs >
 * model_aliases.provider_prefs (alias, openrouter-chat) > global env pin.
 */

export const PROVIDER_QUANTIZATIONS = [
  "int4",
  "int8",
  "fp4",
  "mxfp4",
  "nvfp4",
  "fp6",
  "fp8",
  "mxfp8",
  "fp16",
  "bf16",
  "fp32",
  "unknown",
] as const;
export type ProviderQuantization = (typeof PROVIDER_QUANTIZATIONS)[number];

export const PROVIDER_SORT_BY = ["price", "throughput", "latency"] as const;
export type ProviderSortBy = (typeof PROVIDER_SORT_BY)[number];

export const PROVIDER_PERCENTILES = ["p50", "p75", "p90", "p99"] as const;
export type ProviderPercentile = (typeof PROVIDER_PERCENTILES)[number];

export type ProviderThreshold = number | Partial<Record<ProviderPercentile, number>>;

/** The `provider` object as stored on the gateway and sent to OpenRouter. */
export interface ProviderPrefs {
  only?: string[];
  order?: string[];
  ignore?: string[];
  allow_fallbacks?: boolean;
  require_parameters?: boolean;
  data_collection?: "allow" | "deny";
  zdr?: boolean;
  enforce_distillable_text?: boolean;
  quantizations?: ProviderQuantization[];
  sort?: ProviderSortBy | { by: ProviderSortBy; partition?: "model" | "none" };
  max_price?: { prompt?: number; completion?: number; request?: number; image?: number };
  preferred_min_throughput?: ProviderThreshold;
  preferred_max_latency?: ProviderThreshold;
}

const KNOWN_KEYS = new Set<keyof ProviderPrefs>([
  "only",
  "order",
  "ignore",
  "allow_fallbacks",
  "require_parameters",
  "data_collection",
  "zdr",
  "enforce_distillable_text",
  "quantizations",
  "sort",
  "max_price",
  "preferred_min_throughput",
  "preferred_max_latency",
]);

export const PROVIDER_PREFS_MAX_BYTES = 4096;

class PrefsError extends Error {}

function fail(msg: string): never {
  throw new PrefsError(`provider_prefs: ${msg}`);
}

function slugList(name: string, v: unknown, allowed?: readonly string[]): string[] {
  if (!Array.isArray(v)) fail(`${name} deve ser uma lista`);
  if (v.length === 0) fail(`${name} não pode ser lista vazia (remova o campo)`);
  if (v.length > 32) fail(`${name} tem entradas demais (máx 32)`);
  return v.map((s) => {
    if (typeof s !== "string" || s === "" || /\s/.test(s) || s.length > 64) {
      fail(`${name}: entrada inválida ${JSON.stringify(s)}`);
    }
    if (allowed && !allowed.includes(s)) {
      fail(`${name}: ${JSON.stringify(s)} não é um de ${allowed.join(",")}`);
    }
    return s;
  });
}

function nonNeg(name: string, v: unknown): number {
  if (typeof v !== "number" || !Number.isFinite(v) || v < 0) {
    fail(`${name} deve ser número >= 0`);
  }
  return v;
}

function bool(name: string, v: unknown): boolean {
  if (typeof v !== "boolean") fail(`${name} deve ser true/false`);
  return v;
}

function threshold(name: string, v: unknown): ProviderThreshold {
  if (typeof v === "number") return nonNeg(name, v);
  if (v && typeof v === "object" && !Array.isArray(v)) {
    const entries = Object.entries(v as Record<string, unknown>);
    if (entries.length === 0) fail(`${name} não pode ser objeto vazio`);
    const out: Partial<Record<ProviderPercentile, number>> = {};
    for (const [k, val] of entries) {
      if (!(PROVIDER_PERCENTILES as readonly string[]).includes(k)) {
        fail(`${name}: percentil desconhecido ${JSON.stringify(k)} (use p50/p75/p90/p99)`);
      }
      out[k as ProviderPercentile] = nonNeg(`${name}.${k}`, val);
    }
    return out;
  }
  fail(`${name} deve ser número ou {p50|p75|p90|p99: número}`);
}

/**
 * Validate + normalize an arbitrary value into a `ProviderPrefs`. Throws an
 * Error with a pt-BR message on the first violation. Returns `null` for
 * `null`/`undefined` (= clear). An empty object is rejected.
 */
export function validateProviderPrefs(input: unknown): ProviderPrefs | null {
  if (input === null || input === undefined) return null;
  if (typeof input !== "object" || Array.isArray(input)) fail("deve ser um objeto JSON");
  const raw = input as Record<string, unknown>;
  const out: ProviderPrefs = {};
  for (const key of Object.keys(raw)) {
    if (!KNOWN_KEYS.has(key as keyof ProviderPrefs)) fail(`campo desconhecido ${JSON.stringify(key)}`);
    const v = raw[key];
    if (v === undefined || v === null) continue;
    switch (key as keyof ProviderPrefs) {
      case "only":
      case "order":
      case "ignore":
        out[key as "only" | "order" | "ignore"] = slugList(key, v);
        break;
      case "quantizations":
        out.quantizations = slugList(key, v, PROVIDER_QUANTIZATIONS) as ProviderQuantization[];
        break;
      case "allow_fallbacks":
      case "require_parameters":
      case "zdr":
      case "enforce_distillable_text":
        out[key as "allow_fallbacks" | "require_parameters" | "zdr" | "enforce_distillable_text"] =
          bool(key, v);
        break;
      case "data_collection":
        if (v !== "allow" && v !== "deny") fail('data_collection deve ser "allow" ou "deny"');
        out.data_collection = v;
        break;
      case "sort": {
        if (typeof v === "string") {
          if (!(PROVIDER_SORT_BY as readonly string[]).includes(v)) {
            fail("sort deve ser price|throughput|latency");
          }
          out.sort = v as ProviderSortBy;
        } else if (v && typeof v === "object" && !Array.isArray(v)) {
          const o = v as Record<string, unknown>;
          for (const k of Object.keys(o)) {
            if (k !== "by" && k !== "partition") fail(`sort: campo desconhecido ${JSON.stringify(k)}`);
          }
          if (!(PROVIDER_SORT_BY as readonly string[]).includes(o.by as string)) {
            fail("sort.by deve ser price|throughput|latency");
          }
          const sort: { by: ProviderSortBy; partition?: "model" | "none" } = {
            by: o.by as ProviderSortBy,
          };
          if (o.partition !== undefined) {
            if (o.partition !== "model" && o.partition !== "none") fail("sort.partition deve ser model|none");
            sort.partition = o.partition;
          }
          out.sort = sort;
        } else {
          fail("sort inválido");
        }
        break;
      }
      case "max_price": {
        if (!v || typeof v !== "object" || Array.isArray(v)) fail("max_price deve ser objeto");
        const o = v as Record<string, unknown>;
        const mp: NonNullable<ProviderPrefs["max_price"]> = {};
        for (const k of Object.keys(o)) {
          if (!["prompt", "completion", "request", "image"].includes(k)) {
            fail(`max_price: campo desconhecido ${JSON.stringify(k)}`);
          }
          if (o[k] === undefined || o[k] === null) continue;
          mp[k as keyof typeof mp] = nonNeg(`max_price.${k}`, o[k]);
        }
        if (Object.keys(mp).length === 0) fail("max_price precisa de prompt/completion/request/image");
        out.max_price = mp;
        break;
      }
      case "preferred_min_throughput":
      case "preferred_max_latency":
        out[key as "preferred_min_throughput" | "preferred_max_latency"] = threshold(key, v);
        break;
    }
  }
  if (Object.keys(out).length === 0) fail("defina ao menos um campo (ou limpe para remover)");
  const size = JSON.stringify(out).length;
  if (size > PROVIDER_PREFS_MAX_BYTES) fail(`excede ${PROVIDER_PREFS_MAX_BYTES} bytes`);
  return out;
}

/** Short human summary for tables ("only: novita,deepinfra · zdr · ≤$0.14/$0.50"). */
export function summarizeProviderPrefs(p: ProviderPrefs | null | undefined): string {
  if (!p) return "—";
  const parts: string[] = [];
  if (p.only) parts.push(`only: ${p.only.join(",")}`);
  if (p.order) parts.push(`order: ${p.order.join(",")}`);
  if (p.ignore) parts.push(`ignore: ${p.ignore.join(",")}`);
  if (p.allow_fallbacks === false) parts.push("sem fallback");
  if (p.require_parameters) parts.push("require_params");
  if (p.data_collection === "deny") parts.push("no-collect");
  if (p.zdr) parts.push("zdr");
  if (p.quantizations) parts.push(`quant: ${p.quantizations.join(",")}`);
  if (p.sort) parts.push(`sort: ${typeof p.sort === "string" ? p.sort : p.sort.by}`);
  if (p.max_price) {
    const mp = p.max_price;
    parts.push(`≤$${mp.prompt ?? "∞"}/${mp.completion ?? "∞"}`);
  }
  if (p.preferred_min_throughput !== undefined) parts.push("min tok/s");
  if (p.preferred_max_latency !== undefined) parts.push("max lat");
  return parts.length ? parts.join(" · ") : "—";
}
